src/gmpints.c

```diff
@@ -238,38 +250,135 @@ void LoadInt( Obj gmp )
 
 /****************************************************************************
 **
-*F  NEW_INT( <gmp> )
+**  In order to use the high-level GMP mpz_* functions conveniently while
+**  retaining a low overhead, we introduce fake_mpz_t, which can be thought
+**  of as a "subclass" of mpz_t, extending it by temporary storage for
+**  single limb of data, as well as a reference to a corresponding GAP
+**  object.
+**
+**  For an example on how to correctly use this, see GcdInt.
+*/
+typedef struct {
+  mpz_t v;
+  mp_limb_t tmp;
+  Obj obj;
+} fake_mpz_t[1];
+
+
+/****************************************************************************
+**
+*F  NEW_FAKEMPZ( <fake>, <size> )
 **
+**  Setup a fake mpz_t object for capturing the output of a GMP mpz_* function,
+**  with space for up to <size> limbs allocated.
 **
 */
-static inline Obj NEW_INT( Obj gmp )
+static void NEW_FAKEMPZ( fake_mpz_t fake, UInt size )
 {
-  Obj new;
+  fake->v->_mp_alloc = size;
+  fake->v->_mp_size = 0;
+  if (size == 1) {
+    fake->obj = 0;
+  }
+  else {
+    fake->obj = NewBag( T_INTPOS, size * sizeof(mp_limb_t) );
+  }
+}
 
-  new = NewBag( TNUM_OBJ(gmp), SIZE_OBJ(gmp) );
-  memcpy( ADDR_INT(new), ADDR_INT(gmp), SIZE_OBJ(gmp) );
 
-  return new;
+/****************************************************************************
+**
+*F  FAKEMPZ_GMPorINTOBJ( <fake>, <op> )
+**
+**  Initialize <fake> to reference the content of <op>. For this, <op>
+**  must be an integer, either small or large, but this is *not* checked.
+**  The calling code is responsible for any verification.
+**
+**  WARNING: Since this may trigger a garbage collection, make sure to
+**  call this *before* any calls to FAKEMPZ_GMPorINTOBJ.
```

Comment:
Good catch, that is an obsolete comment I forget to remove (and somehow it indeed migrated to the "wrong" function -- I think used to be a comment for `NEW_FAKEMPZ`, or rather its predecessor). With this final implementation, the GC problems are mostly avoided by using the `MPZ_FAKEMPZ` macro.